### PR TITLE
Allow the column block in the inserter

### LINKED ISSUE #19354: Columns block: No blocks found when using Top Toolbar Inserter while Column block is selected
**Describe the bug**
When I have a single Column block selected within the Columns block and then click the `+` icon in the Top Toolbar, the Block Library shows "No blocks found".  This is super confusing.

**To reproduce**
1. Add a Columns block to the page.
2. Make sure the Column block is selected.
3. Immediately click the `+` icon in the Top Toolbar.
4. Observe the "no blocks found" messaging.

**Expected behavior**
I originally expected a list of blocks that can be added, but was completely confused by the messaging and didn't understand why. It wasn't until I realized that I had the single Column block selected. Maybe we should display better messaging in the Inserter in this case?

**Screenshots**

cols 2019-12-27 14_30_36

**Desktop (please complete the following information):**
 - Mac OS Catalina
 - Firefox 71.0
 - Gutenberg 7.1

### PR DESCRIPTION
**Pull request by youknowriad**
closes #19354 

Since the "column" block already has a "parent" property defined, we should just let it appear in the inserter if a sibling is selected, allowing you to insert another column to the columns block without using the "columns count" range.

Thoughts?

### CODE REVIEW

**Comment by aduth** [comment]
At a glance, it seems reasonable to me. It starts to beg the question whether we even want the "Columns" slider at all, if we're starting to go down this path of letting users rearrange, add, and remove columns in-place.

**Comment by youknowriad** [comment]
Yeah I had the same though but adding columns with the inserter is still a bit painful (you need to have the "columns" or "column" selected. 

There's also something else, the range is limited to 6 columns I believe, do we need a "maxInnerBlocks" API or something? or is it ok to keep it as is. People can still add an illimited number in HTML.

**Comment by aduth** [comment]
That's an interesting point to highlight. It feels like something where, if we feel the need to have some restriction, we would want to allow blocks supports to be changed dynamically†. So, for example, the Columns block could dynamically set `inserter: false` when the maximum number of columns is reached. I don't know that it's common enough a use-case that we would want to optimize some generic solution (like `maxInnerBlocks`). Even for the existing Columns case, it's a pretty questionable restriction (and not enforced in any way at the markup level).

_† Maybe it already is? Is there something to dispatch to the `core/blocks` store to change these supports after the initial registration?_

**Comment by youknowriad** [comment]
> † Maybe it already is? Is there something to dispatch to the core/blocks store to change these supports after the initial registration?

I'm pretty sure it's possible but we don't have dedicated actions.

**Comment by youknowriad** [comment]
Actually, no, it's not possible because we only want to do this change of a single instance.

**Comment by aduth** [comment]
Ah, I see how that's an issue. I could possibly imagine some hierarchy supporting optional per-instance block supports, but aside from this use-case, I'm not sure how useful that would be.

**Comment by aduth** [comment]
I tested this and it feels a little strange to me. Intuitively, I felt that I would _add a Column to a Columns block_. Accordingly, I selected the Columns block and clicked the inserter, but was presented with all block options, and not the Column block. Reflecting on this and how the inserter works in general (you're inserting relative the selected block in the context of the parent or root), I understand the technical reasons for why this is. But I wonder if my experiences will be a common source of confusion.

On the other hand, it could be argued that even if it's not optimized to insert this way, there will always be an inserter for when the Column is selected, so it may as well present something useful, rather than "No blocks found" (#19354). In that sense, I think it can be an okay change.

**Review by aduth — APPROVED** [review]
Code-wise, this looks correct. I'll leave it to your discretion on whether to proceed, given the thought process I shared at https://github.com/WordPress/gutenberg/pull/20502#issuecomment-592092143.

**Comment by youknowriad** [comment]
> But I wonder if my experiences will be a common source of confusion.

I agree with this confusion and I had a similar feeling. The inserter works consistently no matter which block is selected: "insert after the block". We could potentially consider changing that into:

 - If the block supports inner blocks: insert inside the block
 - otherwise, insert after

Even with this, there's a potential for confusion though.

**Comment by jbinda** [comment]
Together with @koke, @pinarol and @iamthomasbishop we were discussing how to approach Columns block on mobile. This PR help us a lot to handle Inserter behaviour. However I also wonder about other two related topics.

1. How does it affects `Column` number restriction. Currently we can set  `Column` number between 2 and 6 via Slider in Sidebar. I think adding `Column` via Inserter gives possibility to pass that limitation without further changes around Inserter logic.
2. Shouldn't we also add remove option for `Column` in dropdown menu in BlockControls after block gets selection ? It's about being consistent - if you are able to add block via Inserter probably you would like to remove it as well in the same way as other blocks

What do you think about this ?

**Comment by youknowriad** [comment]
> if you are able to add block via Inserter probably you would like to remove it as well in the same way as other blocks

This is already the case.

**Comment by jbinda** [comment]
Ok, great ! Thanks for answer

**Comment by koke** [comment]
If we want to limit the column number to the 2-6 range, then any sort of direct manipulation makes the UX confusing. We could allow inserting columns and have that new API to limit the number of child blocks, but then it would not be clear to a user why sometimes they can insert and sometimes they can't. Similarly, if there are two remaining columns, it wouldn't be clear why they can't remove columns now. The slider might not be the most obvious way to manipulate the column number, but at least makes these limits explicit.

I'm wondering if there's a way to relax that restriction, making the 2-6 range more of a suggestion than a limit. That way, you would be able to insert as many columns as you like, and remove them all, leaving an empty columns block that would show the initial placeholder.

> I don't know that it's common enough a use-case that we would want to optimize some generic solution (like maxInnerBlocks).

We came across another potential use case in https://github.com/wordpress-mobile/gutenberg-mobile/issues/1607#issuecomment-596407390, where it would be interesting to refactor media-text so it uses inner blocks for the media part. I believe the current API doesn't allow both limiting to either image or video, and a maximum of one child block. You can either lock the template to an explicit block, or allow multiple children.

**Comment by jbinda** [comment]
>If we want to limit the column number to the 2-6 range, then any sort of direct manipulation makes the UX confusing

It seems that the max limitation is not necessary - it seems you can put as many Column as you want. Avoiding minimum Column limitation is more tricky because at first glance it was difficult to me to think about Columns with only one Column inside. However second thought was that there might be some use case and user would like to have Columns block with only one Column so why we should block it ? 

If we're going to remove the limitation I also wonder do we still want to show slider in Sidebar settings ?

**Comment by youknowriad** [comment]
Added a warning message if the columns count exceeds the maximum. What do you think @aduth ? I believe it's a bit better than trying to force a columns count.

**Comment by aduth** [comment]
Taking a fresh look at this, I find the previously-noted concerns interesting to consider. I agree that it feels odd to allow for this sort of direct manipulation where a "cap" exists (even if it's now a soft-cap).

For me, I think a cap could exist, but the current options for giving feedback to the user are non-ideal. I think the ideal case for me would be one where maybe the inserter option becomes grayed out with an inlined or tooltip message about how the maximum number of columns has been reached. It's obviously a bigger task than we might want to take on here, but it could be a useful inserter feature generally for blocks to leverage.

The notice is okay as far as providing feedback, and there's some prior art here with things like the Contrast Checker (give a user the freedom to make a poor choice, but inform them about it). But it could be strange that we only inform a user of a mistake after the fact (could be missed). I guess it comes down to a question of "How bad of a problem is it?". It may not really be that big of a problem for there to be many columns. Or at least, the user can use their own judgment to decide based on how it appears on their site.

I think the options I see here are one of:

1. Abandon the pull request, considering these usability concerns. 
2. Keep the notice.
3. Consider it blocked by some other effort to expose interfaces for "block insertion restrictions".

Might also be good to get some design/usability feedback as well.

File: packages/block-library/src/columns/edit.js
Comment (aduth) [inline packages/block-library/src/columns/edit.js:86]: The use of "big" here as it relates to the number is somewhat confusing, since it could be misinterpreted as "the columns themselves occupy more visual space".

Maybe something like:

```suggestion
								'This column count exceeds the recommended amount and may cause visual breakage.'
```

**Comment by youknowriad** [comment]
> Abandon the pull request, considering these usability concerns.

One of the reason I don't want to abandon the PR is because it will allow the inbetween inserter to trigger to add columns when we introduce the horizontal alignment support. I believe @ellatrix is working on some of that.

**Comment by ellatrix** [comment]
Yes, I've been thinking it could be useful to have the block list appender visible for columns. Not sure what's the best conditions for it to be visible. When the last column is selected? When any block is selected? When the inner block list is selected as well? We also need to think about this together with other blocks like the navigation block, social icons, buttons... which should have the UI.

**Comment by youknowriad** [comment]
Let's give this a try, it's easy to revert anyway.

**Pull request merged**
